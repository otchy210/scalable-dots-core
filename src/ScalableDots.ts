import { SvgDocument } from './SvgDocument';
import { ScalableDotsProps } from './types';
import { uint8toHex, unit8toRatio } from './utils';

class ScalableDots {
  private svgDocument;
  constructor({ imageData, type, size, gap }: ScalableDotsProps) {
    const width = (size + gap) * imageData.width + gap;
    const height = (size + gap) * imageData.height + gap;
    this.svgDocument = new SvgDocument({ width, height });
    for (let y = 0; y < imageData.height; y++) {
      for (let x = 0; x < imageData.width; x++) {
        const offset = (y * imageData.width + x) * 4;
        const [r, g, b, a] = imageData.data.slice(offset, offset + 4);
        if (a === 0) {
          continue;
        }
        const fill =
          a === 255
            ? `#${uint8toHex(r)}${uint8toHex(g)}${uint8toHex(b)}`
            : `rgba(${r},${g},${b},${unit8toRatio(a)})`;
        const dotProps = {
          x: x * (size + gap),
          y: y * (size + gap),
          size,
          fill,
        };
        this.svgDocument.addDot(type, dotProps);
      }
    }
  }
  toMinifiedXml() {
    return this.svgDocument.toMinifiedXml();
  }
  toPrettyXml(tab = '  ') {
    return this.svgDocument.toPrettyXml(tab);
  }
}

export const scalableDots = (props: ScalableDotsProps) => {
  return new ScalableDots(props);
};
