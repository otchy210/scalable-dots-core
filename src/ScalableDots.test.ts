import { scalableDots } from './ScalableDots';

describe('scalebleDots', () => {
  it('works', () => {
    const data = new Uint8ClampedArray([
      255, 0, 0, 255, 0, 192, 0, 255, 0, 0, 255, 255, 192, 192, 0, 255,
    ]);
    const imageData = { width: 2, height: 2, data };

    const dots = scalableDots({ imageData, type: 'SQUARE', size: 16, gap: 2 });
    expect(dots.toMinifiedXml()).toBe(
      '<svg viewBox="0 0 38 38" xmlns="http://www.w3.org/2000/svg" version="1.1"><path d="m0 0h16v16h-16z" fill="#ff0000"/><path d="m18 0h16v16h-16z" fill="#00c000"/><path d="m0 18h16v16h-16z" fill="#0000ff"/><path d="m18 18h16v16h-16z" fill="#c0c000"/></svg>'
    );
  });
});
