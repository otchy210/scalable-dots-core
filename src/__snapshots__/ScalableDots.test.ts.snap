// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scalebleDots renders RGB case properly 1`] = `
"<svg viewBox=\\"0 0 34 34\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\">
  <path d=\\"m0 0h16v16h-16z\\" fill=\\"#ff0000\\"/>
  <path d=\\"m18 0h16v16h-16z\\" fill=\\"#00c000\\"/>
  <path d=\\"m0 18h16v16h-16z\\" fill=\\"#0000ff\\"/>
  <path d=\\"m18 18h16v16h-16z\\" fill=\\"#c0c000\\"/>
</svg>"
`;

exports[`scalebleDots renders extream options case properly 1`] = `
"<svg viewBox=\\"0 0 18 18\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\">
  <path d=\\"m0 0h1v1h-1z\\" fill=\\"#ff0000\\"/>
  <path d=\\"m17 0h1v1h-1z\\" fill=\\"#00c000\\"/>
  <path d=\\"m0 17h1v1h-1z\\" fill=\\"#0000ff\\"/>
  <path d=\\"m17 17h1v1h-1z\\" fill=\\"#c0c000\\"/>
</svg>"
`;

exports[`scalebleDots renders half-transparent case properly 1`] = `
"<svg viewBox=\\"0 0 34 34\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\">
  <path d=\\"m0 0h16v16h-16z\\" fill=\\"rgba(255,0,0,.5)\\"/>
  <path d=\\"m18 0h16v16h-16z\\" fill=\\"rgba(0,192,0,.5)\\"/>
  <path d=\\"m0 18h16v16h-16z\\" fill=\\"rgba(0,0,255,.5)\\"/>
  <path d=\\"m18 18h16v16h-16z\\" fill=\\"rgba(192,192,0,.5)\\"/>
</svg>"
`;

exports[`scalebleDots renders transparent case properly 1`] = `
"<svg viewBox=\\"0 0 34 34\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\">
  <path d=\\"m18 0h16v16h-16z\\" fill=\\"rgba(0,192,0,.5)\\"/>
  <path d=\\"m0 18h16v16h-16z\\" fill=\\"#0000ff\\"/>
</svg>"
`;
